// @ethereumjs/block/dist/header-from-rpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const header_1 = require("./header");
const helpers_1 = require("./helpers");
/**
 * Creates a new block header object from Ethereum JSON RPC.
 *
 * @param blockParams - Ethereum JSON RPC of block (eth_getBlockByNumber)
 * @param chainOptions - An object describing the blockchain
 */
function blockHeaderFromRpc(blockParams, options) {
    const { parentHash, sha3Uncles, miner, stateRoot, transactionsRoot, receiptRoot, receiptsRoot, logsBloom, difficulty, number, gasLimit, gasUsed, timestamp, extraData, mixHash, nonce, } = blockParams;
    let baseFeePerGas;
    // Check if the field baseFeePerGas is present in the block.
    // This field was introduced after: https://eips.ethereum.org/EIPS/eip-1559
    if ('baseFeePerGas' in blockParams) {
        baseFeePerGas = blockParams.baseFeePerGas;
    }
    const blockHeader = header_1.BlockHeader.fromHeaderData({
        parentHash,
        uncleHash: sha3Uncles,
        coinbase: miner,
        stateRoot,
        transactionsTrie: transactionsRoot,
        receiptTrie: receiptRoot || receiptsRoot,
        bloom: logsBloom,
        difficulty: helpers_1.numberToHex(difficulty),
        number,
        gasLimit,
        gasUsed,
        timestamp,
        extraData,
        mixHash,
        nonce,
        baseFeePerGas,
    }, options);
    return blockHeader;
}
exports.default = blockHeaderFromRpc;
